rust-runtime/smithy-json/src/serialize.rs

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use crate::escape::escape_string;
+
+pub struct JsonObjectWriter<'a> {
+    json: &'a mut String,
+    started: bool,
+}
+
+impl<'a> JsonObjectWriter<'a> {
+    pub fn new(output: &'a mut String) -> Self {
+        output.push('{');
+        Self {
+            json: output,
+            started: false,
+        }
+    }
+
+    /// Writes a null value with the given `key`.
+    pub fn null(&mut self, key: &str) -> &mut Self {
+        self.key(key);
+        self.json.push_str("null");
+        self
+    }
+
+    /// Writes the boolean `value` with the given `key`.
+    pub fn boolean(&mut self, key: &str, value: bool) -> &mut Self {
+        self.key(key);
+        self.json.push_str(match value {
+            true => "true",
+            _ => "false",
+        });
+        self
+    }
+
+    /// Writes a string `value` with the given `key`.
+    pub fn string(&mut self, key: &str, value: &str) -> &mut Self {
+        self.key(key);
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a number `value` (already converted to string) with the given `key`.
+    pub fn number(&mut self, key: &str, value: &str) -> &mut Self {
```

Comment:
I will leave this up to you, but I think we will probably want this library to handle numeric JSON conversion: https://github.com/awslabs/smithy-rs/blob/42c1f215a5f88cd58871117dee3a67fa5b0b318f/rust-runtime/smithy-types/src/lib.rs#L57

Otherwise, we'll need to handle it upstream in a bunch of different places

Reply:
Oh, I wasn't aware of the `Number` type in `smithy-types`. That simplifies things considerably. Originally, I was thinking I would need to pull in the `num` crate in order to write a generic number function, or alternatively make a separate number function for each number type and have the code generator choose, but I settled for just calling `to_string()` everywhere this `number` function is called.

Will refactor.

Reply:
Also moved the timestamp handling code out of the code gen and into the serializers so that it's not repeated everywhere.